app/src/main/java/me/hufman/androidautoidrive/phoneui/viewmodels/CarDrivingStatsModel.kt

```diff
@@ -1,252 +1,454 @@
+
+	val averageSpeed2 = carInfo.cachedCdsData.liveData[CDS.DRIVING.AVERAGESPEED].map {
+		it.tryAsJsonObject("averageSpeed")?.tryAsJsonPrimitive("averageSpeed2")?.tryAsDouble?.takeIf { it < 2093 }
+	}.format("%.1f").addUnit(unitsAverageSpeedLabel)
+
+	val drivingStyleAccel = carInfo.cachedCdsData.liveData[CDS.DRIVING.DRIVINGSTYLE].map {
+		it.tryAsJsonObject("drivingStyle")?.tryAsJsonPrimitive("accelerate")?.tryAsInt
+	}
+	val drivingStyleBrake = carInfo.cachedCdsData.liveData[CDS.DRIVING.DRIVINGSTYLE].map {
+		it.tryAsJsonObject("drivingStyle")?.tryAsJsonPrimitive("brake")?.tryAsInt
+	}
+	val drivingStyleShift = carInfo.cachedCdsData.liveData[CDS.DRIVING.DRIVINGSTYLE].map {
+		it.tryAsJsonObject("drivingStyle")?.tryAsJsonPrimitive("shift")?.tryAsInt
+	}
+
+	val ecoRangeWon = carInfo.cachedCdsData.liveData[CDS.DRIVING.ECORANGEWON].map {
+		it.tryAsJsonPrimitive("ecoRangeWon")?.tryAsDouble
+	}.combine(units) { value, units ->
+		units.distanceUnits.fromCarUnit(value)
+	}.format("%.1f").addUnit(unitsDistanceLabel)
+
+	fun update() {
+		// any other updates
+		_idriveVersion.value = carInfo.capabilities["hmi.version"]
+	}
+
+	/* JEZIKK additions */
+	val engineTemp = carInfo.cachedCdsData.liveData[CDS.ENGINE.TEMPERATURE].map {
+		it.tryAsJsonObject("temperature")?.tryAsJsonPrimitive("engine")?.tryAsDouble
+	}.format("%.0f").addUnit(unitsTemperatureLabel)
+	val oilTemp = carInfo.cachedCdsData.liveData[CDS.ENGINE.TEMPERATURE].map {
+		it.tryAsJsonObject("temperature")?.tryAsJsonPrimitive("oil")?.tryAsDouble
+	}.format("%.0f").addUnit(unitsTemperatureLabel)
+	val speedActual = carInfo.cdsData.liveData[CDS.DRIVING.SPEEDACTUAL].map {
+		it.tryAsJsonPrimitive("speedActual")?.tryAsDouble
+	}.format("%.0f") //.addUnit(unitsAverageSpeedLabel)
+	val speedDisplayed = carInfo.cdsData.liveData[CDS.DRIVING.SPEEDDISPLAYED].map {
+		it.tryAsJsonPrimitive("speedDisplayed")?.tryAsDouble
+	}.format("%.0f").addUnit(unitsAverageSpeedLabel)
+	val tempInterior = carInfo.cdsData.liveData[CDS.SENSORS.TEMPERATUREINTERIOR].map {
+		it.tryAsJsonPrimitive("temperatureInterior")?.tryAsDouble
+	}.format("%.1f").addUnit(unitsTemperatureLabel)
+	val tempExterior = carInfo.cdsData.liveData[CDS.SENSORS.TEMPERATUREEXTERIOR].map {
+		it.tryAsJsonPrimitive("temperatureExterior")?.tryAsDouble
+	}.format("%.1f").addUnit(unitsTemperatureLabel)
+
+	/*
+		0 - Initialisierung -> initialization
+		1 - Tractionmodus -> Traction mode
+		2 - Komfortmodus -> Comfort mode
+		3 - Basismodus -> Basic mode
+		4 - Sportmodus -> Sports mode
+		5 - Sportmodusplus -> Sport+
+		6 - Racemodus -> Race
+		7 - Ecopro ->EcoPro
+		8 - Ecoproplus -> EcoPro+
+		9 - Komfortmoduserweitert -> Comfort mode extended
+		xx - Adaptive Mode
+		15 - Unknown
+
+		Some cars have "+" modes (sport+, comfort+) and some other the same modes are called "Individual".
+	 */
+	//val drivingMode = carInfo.cdsData.liveData[CDS.DRIVING.MODE].map {
+	val drivingMode = carInfo.cachedCdsData.liveData[CDS.DRIVING.MODE].map {

```

Comment:
Driving Mode doesn't need to be from cachedCdsData, I think, so it could be pulled from `cdsData` and it can be cleaned from the CACHED_CDS_KEYS above.

Reply:
done.